DevOps Principles - Draft Ready for Review. (#5)
SkillsFundingAgency/das-technical-guidance: principles/devops.md

# DevOps

## Introduction

**DevOps is a service wide way of working, we are all in this together.** It is not the tools, nor is it a single team. It is famously summarised in a single sentance by Donovan Brown (Principal DevOps Manager at Microsoft):

> DevOps is the union of people, process, and products to enable continuous delivery of value to our end users.

> We follow this ethos to ensure delivery of value to all of the Digital Apprenticeship Service users, internal and external. It is a combination of the passion, hard work and process that enables DevOps culture within the service. The continuous set of processes that we follow forms the methodology for our Agile planning and continuous delivery, for example aligned sprints across all teams.

Reviewer: ```suggestion
We follow this ethos to ensure delivery of value to all of the Apprenticeship Service users, internal and external. It is a combination of the passion, hard work and process that enables DevOps culture within the service. The continuous set of processes that we follow forms the methodology for our Agile planning and continuous delivery, for example aligned sprints across all teams.
```

> Adhereing to DevOps provides the service with the following benefits:

Reviewer: ```suggestion
Adhering to DevOps provides the service with the following benefits:
```

- **Agile ways of working.** Across the service we perform daily collaboration through Agile ceremonies; sprint planning, daily standup, sprint reviews and retrospectives. Support and trust are key to our ways of working which helps enable our continuous delivery. There are also other ceremonies that bring all of the teams together such as the Scrum of Scrums, Clan Leads meeting, Architecture board, Quality Guild and a regular service wide standup.

- **Show and tell.** These are regular meetings that allow teams to celebrate their work and talk about what they’ve learned. It's a chance to bring teams together, share success and improve on collaboration. Progress is highlighted along with question and feedback sessions. Transparency is encouraged to glimpse 'behind the curtain' of the products in development and to highlight where dependencies may exist between teams. A great show and tell session is one that is performed on a regular basis to ensure continuous feedback.

## Automation

> This is the most vital factor of the DevOps culture. The purpose of automation is to build a system such that it eliminates various human errors and can fix errors and repeating the successful process again without any sort of intervention or delays. Automation greatly improves the workflow and productivity of the service. 

Reviewer 2: bit of an awkward sentence: `...human errors and can fix errors and repeating the successful process again...`

Author: Updated 👍

> - **Infrastructure as Code (IaC).** IaC is an infrastructure management approach that makes continuous delivery and DevOps possible. Without IaC, our developers and engineers would have to treat each target environment individually, which becomes a tedious task as we have different environments for development, testing, and production use. We include the Azure Resource Manager (ARM) templates and Azure DevOps build definitions along side an applications codebase in the GitHub repository, this is to match the development lifecycle of an application. As application requirements change, so can the build and release along side it. 

Reviewer: ```suggestion
- **Infrastructure as Code (IaC).** IaC is an infrastructure management approach that makes continuous delivery and DevOps possible. Without IaC, our developers and engineers would have to treat each target environment individually, which becomes a tedious task as we have different environments for development, testing, and production use. We include the Azure Resource Manager (ARM) templates and Azure DevOps build definitions alongside an application's codebase in the GitHub repository, this is to match the development lifecycle of an application. As application requirements change, so can the build and release alongside it. 
```

> - **Infrastructure as Code (IaC).** IaC is an infrastructure management approach that makes continuous delivery and DevOps possible. Without IaC, our developers and engineers would have to treat each target environment individually, which becomes a tedious task as we have different environments for development, testing, and production use. We include the Azure Resource Manager (ARM) templates and Azure DevOps build definitions along side each an applications repository in GitHub, this is to match the development lifecycle of an application. As application requirements change, so can the build and release along side it. 

Reviewer 2: bit of grammar: `...along side each an applications repository...`

Author: Sorted 👍

> - **Cloud infrastructure.** We are a cloud first service as it provides the flexibility, toolsets, and scalability required. All our architectures rely on PaaS solutions. Through IaC and continuous integration and continuous deployment (CI/CD) we can quickly deploy infrastructure and ensure there is no environment or configuration drift. Builds can be deployed to each environment providing consistant dev/test and production environments. It enables our teams to test applications in production-like environments early in the development cycle.

Reviewer: ```suggestion
- **Cloud infrastructure.** We are a cloud first service as it provides the flexibility, toolsets, and scalability required. All our architectures rely on PaaS solutions. Through IaC and continuous integration and continuous deployment (CI/CD) we can quickly deploy infrastructure and ensure there is no environment or configuration drift. Builds can be deployed to each environment providing consistent dev/test and production environments. It enables our teams to test applications in production-like environments early in the development cycle.
```
